src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java

```diff
@@ -0,0 +1,280 @@
+import org.jenkinsci.plugins.github.util.misc.NullSafeFunction;
+import org.jenkinsci.plugins.github.util.misc.NullSafePredicate;
+import org.jenkinsci.plugins.plaincredentials.StringCredentials;
+import org.jenkinsci.plugins.plaincredentials.impl.StringCredentialsImpl;
+import org.kohsuke.github.GitHub;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+
+import static com.cloudbees.plugins.credentials.CredentialsMatchers.firstOrDefault;
+import static com.cloudbees.plugins.credentials.CredentialsMatchers.withId;
+import static com.cloudbees.plugins.credentials.CredentialsProvider.lookupCredentials;
+import static com.cloudbees.plugins.credentials.domains.URIRequirementBuilder.fromUri;
+import static org.apache.commons.lang3.StringUtils.defaultIfBlank;
+import static org.apache.commons.lang3.StringUtils.defaultIfEmpty;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+/**
+ * This object represents configuration of each credentials-github pair.
+ * If no api url explicitly defined, default url used.
+ * So one github server can be used with many creds and one token can be used multiply times in lot of gh servers
+ *
+ * @author lanwen (Merkushev Kirill)
+ * @since TODO
+ */
+@XStreamAlias("github-server-config")
+public class GitHubServerConfig extends AbstractDescribableImpl<GitHubServerConfig> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GitHubServerConfig.class);
+
+    /**
+     * Because of {@link GitHub} hide this const from external use we need to store it here
+     */
+    public static final String GITHUB_URL = "https://api.github.com";
+
+    /**
+     * Used as default token value if no any creds found by given credsId.
+     */
+    private static final String UNKNOWN_TOKEN = "UNKNOWN_TOKEN";
+
+    private String apiUrl;
+    private boolean manageHooks = true;
+    private final String credentialsId;
+
+    /**
+     * only to set to default apiUrl when uncheck
+     */
+    private transient boolean custom;
+
+    @DataBoundConstructor
+    public GitHubServerConfig(String credentialsId) {
+        this.credentialsId = credentialsId;
+    }
+
+    /**
+     * {@link #custom} field should be defined earlier. Because of we get full content of optional block,
+     * even if it already unchecked. So if we want to return api url to default value - custom value should affect
+     *
+     * @param apiUrl custom url if GH. Set api url to default value if custom is unchecked or value is blank
+     */
+    @DataBoundSetter
+    public void setApiUrl(String apiUrl) {
+        if (custom) {
+            this.apiUrl = defaultIfBlank(apiUrl, GITHUB_URL);
```

Comment:
remove this magic and directly store apiUrl. If use forgot url, then it can add new ServerConfig where default will appear again.